system/lib/wasmfs/file.h

```diff
@@ -66,45 +65,9 @@ class File : public std::enable_shared_from_this<File> {
 
   backend_t getBackend() { return backend; }
 
-  class Handle {
-
-  protected:
-    // This mutex is needed when one needs to access access a previously locked
-    // file in the same thread. For example, rename will need to traverse
-    // 2 paths and access the same locked directory twice.
-    // TODO: During benchmarking, test recursive vs normal mutex performance.
-    std::unique_lock<std::recursive_mutex> lock;
-    std::shared_ptr<File> file;
-
-  public:
-    Handle(std::shared_ptr<File> file) : file(file), lock(file->mutex) {}
-    Handle(std::shared_ptr<File> file, std::defer_lock_t)
-      : file(file), lock(file->mutex, std::defer_lock) {}
-    bool trylock() { return lock.try_lock(); }
-    size_t getSize() { return file->getSize(); }
-    mode_t& mode() { return file->mode; }
-    time_t& ctime() { return file->ctime; }
-    time_t& mtime() { return file->mtime; }
-    time_t& atime() { return file->atime; }
-
-    // Note: parent.lock() creates a new shared_ptr to the same Directory
-    // specified by the parent weak_ptr.
-    std::shared_ptr<File> getParent() { return file->parent.lock(); }
-    void setParent(std::shared_ptr<File> parent) { file->parent = parent; }
-
-    std::shared_ptr<File> unlocked() { return file; }
-  };
-
-  Handle locked() { return Handle(shared_from_this()); }
-
-  std::optional<Handle> maybeLocked() {
-    auto handle = Handle(shared_from_this(), std::defer_lock);
-    if (handle.trylock()) {
-      return Handle(shared_from_this());
-    } else {
-      return {};
-    }
-  }
```

Comment:
Moving the `*::Handle` definitions out of their enclosing classes meant that the `*::Handle` types were incomplete at the locations where the `locked` and `maybeLocked` methods were declared, which meant that the implementations of those methods had to move to a different place where the `*::Handle` types were fully defined. I'll play around with it and see if I can get them back into the header.

Reply:
What did we decide here? I'm still pretty worried that if a getter-type function like these is out of a header than it will impact code size and speed negatively.

Reply:
Oh, these are all back in the header, but now they appear after all the class definitions and are marked `inline`.